### PR TITLE
Convert hara.data.diff and related namespaces to CLJC

### PR DESCRIPTION
**Pull request by danielcompton**
I'm wanting to use hara.data.diff from CLJS, so I've converted it and the related namespaces to CLJC. Let me know if there's anything you'd like me to change.

### CODE REVIEW

File: src/hara/common/checks.cljc [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+     :cljs (goog/isBoolean x)))
+
+(defn hash-map?
+  "Returns `true` if `x` implements `clojure.lang.APersistentMap`.
+
+   (hash-map? {})    => true
+   (hash-map? [])    => false"
+  {:added "2.0"}
+  [x]
+  #?(:clj (instance? clojure.lang.APersistentMap x)
```
Comment (danielcompton) [inline src/hara/common/checks.cljc]: Should the instance check be for the interface `clojure.lang.IPersistentMap` instead of `APersistentMap`?

Comment (zcaudate) [inline src/hara/common/checks.cljc]: the check is for APersistentMap. This is to stop defrecords from passing

Comment (danielcompton) [inline src/hara/common/checks.cljc]: Cool! Just wondering.

**Comment by zcaudate** [comment]
Wow. that's a big commit.

This is kind of cool because I know that people have wanted cljs support for a while now. However, I have a bit of an aversion to cljc. so please give me some time to think about this.

There are also other considerations like whether this works with hydrox (tests/documentation) as well as lein-repack. I have a feeling that these may break.

**Comment by danielcompton** [comment]
Hey, I've split this up into multiple commits so you can see what's changing in each. 0a1df07 is the only meaty commit in the bunch. The total diff is large, but a lot of it is just removing trailing whitespace, adding whitespace when reformatting with the Reader Conditional, or adding an extra paren to wrap Reader Conditionals.

You can view the substantial changes at https://github.com/zcaudate/hara/pull/18/commits/0a1df07308b2a4ad0cf4a3782e173100e6bbabd9?w=0.

**Comment by danielcompton** [comment]
If you can give me the commands you use to run hydrox and repack, I can take a look to see what's possible. Two other alternatives:

1. Copy hara.data.diff into a cljs file and remove the parts that don't work in cljs. Do the same for its dependencies.
2. Copy hara.data.diff and related files into my own project so I can work separately with my CLJC files.

Totally understand your reticence, cljc does add a little bit of work, especially for workflow type things, but it's worth it IMO.

**Comment by zcaudate** [comment]
Looking at the PR again, I think that cljs would be best if it were separated into another project.

There's a lot of automation going on with hara. I need it to maintain a codebase that big. 
 - all docstrings are generated from tests
 - the entire website is generated from tests
 - all packages are "repacked" into around 50 different jars and pushed onto maven
 
So any change that breaks this flow severely restricts the ability to work.

What we can have with clj/cljs intergration is to have shared tests and a spec. I'm actually more interested in the spec here https://github.com/zcaudate/hara/issues/15 because it is a way to unify functionality without (in my opinion) butchering code. Like I know it's so tempting to be able to add a `#?{:clj .... }` and have it work but I think it'd be a much better solution if it were pure cljs.

**Comment by zcaudate** [comment]
the commands that i feel may break are:

lein hydrox import
lein repack

**Comment by danielcompton** [comment]
Hey, I've had a bit of a look at hydrox and repack, and I don't think that updating them to support CLJC will be trivial (at least for me with no prior knowledge of the code).

The benefit of cljc is that quite a lot of the code doesn't need to be changed at all, for example `hara.data.{diff,map,nested}` needed virtually no changes to support clj, and cljs, and any fixes for one will be fixed in both. Sharing tests would also be easier with cljc, as you could run the same tests. If you start a cljs sister project then you'll need to backport patches, e.t.c. which sounds messy and error prone.

I think for now we'll just vendor the few files that we need from `hara` into our own app, as I don't have the time to fix hydrox and repack. We may revisit this in the future to merge back in when we've got a few more cycles spare. Sound good?

**Comment by zcaudate** [comment]
Thanks Daniel,

This has been something that is on the back of my mind but it definitely is a big task. 

The first thing that should be done is the unification of tests. They can be in cljc format.

For source code generation I prefer code generation over the reader macro.

 dalap offers a much cleaner/better solution than cljx/cljc - even though cljc is definitely easier to work with on smaller projects.
http://birdseyesoftware.github.io/lein-dalap.docs/ <http://birdseyesoftware.github.io/lein-dalap.docs/>

To tell you the truth, I think cljc is by far, the worst thing that has happened to clojure because it harks back to the bad old days of c/c++ preprocessor macros. They made code horrendously difficult to read because platform specific code started to fragment the codebase.  The interesting thing about clojure is that most people don’t mind because most libraries tend to be less than 1000 loc and so using clix doesn’t really matter. However, hara is different: It’s big, it’s modular and it relies on automation for project maintenance. I find it hard enough reading/reasoning about code, let alone having some preprocessor code for platform specific behaviour be sprinkled throughout the codebase. In my option, the test of good code is readability - and I’m not convinced at all of the advantages that cljc offers to a large code base in the long run. Once fragmentation gets into a codebase, it’s hard to stop the rot and I prefer to stay on the side of vigilance.

**Pull request closed without merging**